### PR TITLE
Remove consistent check in test for job suspend

### LINKED ISSUE #121013: Flaky Test [Sig-Apps]: Job should not create pods when created in suspend state
### Which jobs are flaking?

[triage](https://storage.googleapis.com/k8s-triage/index.html?ci=0&pr=1&sig=apps&test=Job%20should%20not%20create%20pods%20when%20created%20in%20suspend%20state)

In [pull-kubernetes-e2e-kind](https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-kind) and [pull-kubernetes-e2e-kind-ipv6](https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-kind-ipv6), I started seeing these flakes.  

The job in question has not been very flaky before so this is odd.  

### Which tests are flaking?

In [pull-kubernetes-e2e-kind](https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-kind) and [pull-kubernetes-e2e-kind-ipv6](https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-kind-ipv6), I started seeing these flakes.  

The job "Job should not create pods when created in suspend state" is the one that is flaking.

### Since when has it been flaking?

Triage says a few days ago.

### Testgrid link

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-kind

### Reason for failure (if possible)

_No response_

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig apps

### PR DESCRIPTION
**Pull request by kannon92**
#### What type of PR is this?
/kind flake

#### What this PR does / why we need it:
Eliminate potential race condition when checking if the job has Suspended condition. Current implementation is fragile because it takes time for the Job controller to add the condition. On the other hand, the waiting in test happens as a side effect of calling Consistently function which is used to verify pods aren't created for a while.

#### Which issue(s) this PR fixes:

Related to #121013

#### Special notes for your reviewer:
See #121013 for some thoughts.
#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/e2e/apps/job.go
Comment (alculquicondor) [inline test/e2e/apps/job.go:293]: this should probably return the last version of the job to be used below.

File: test/e2e/apps/job.go
Comment (alculquicondor) [inline test/e2e/apps/job.go:290]: We can optimize this test a little:
After we waited for the job to have the suspend field, we can do a one time check that there are no pods (as we know that a sync already happened), instead of using `Consistently`, which is a busy wait.

**Comment by kannon92** [comment]
@alculquicondor do you know how to run these tests locally?

File: test/e2e/apps/job.go
Comment (kannon92) [reply test/e2e/apps/job.go:292]: Pushed a change for that.

File: test/e2e/apps/job.go
Comment (kannon92) [reply test/e2e/apps/job.go]: yea, good call. Fixed.

**Comment by kannon92** [comment]
/hold

Let's wait and see if this will help flakiness.

File: test/e2e/apps/job.go
Comment (alculquicondor) [inline test/e2e/apps/job.go:291]:
```suggestion
		gomega.Expect(pods.Items).To(gomega.HaveLen(0))
```

File: test/e2e/apps/job.go
Comment (alculquicondor) [inline test/e2e/apps/job.go:292]: ExpectNoError should be before the check for the length of the response :)

**Review by alculquicondor — COMMENTED** [review]
I haven't done it in a while, but you can run the tests by creating a kind cluster and using `/hack/ginkgo-e2e.sh`.

**Comment by alculquicondor** [comment]
/retitle Fix flaky test for Job suspend

File: test/e2e/apps/job.go
Comment (alculquicondor) [inline test/e2e/apps/job.go:289]: this looks repeated. Also, we shouldn't need `if err != nil` if we have `framework.ExpectNoError`

Comment (kannon92) [reply test/e2e/apps/job.go]: updated.

File: test/e2e/apps/job.go
Comment (kannon92) [reply test/e2e/apps/job.go]: updated.

File: test/e2e/apps/job.go
Comment (kannon92) [reply test/e2e/apps/job.go]: BTW, you have to use `gomega.BeEmpty()` for testing an empty list.

File: test/e2e/apps/job.go
Comment (kannon92) [reply test/e2e/apps/job.go:283]: I had to revert that change.  We got errors saying that the job was not the latest and that caused a failure downstream in the job.  I removed the return and I retrieve the job before patching it to unsuspend.

**Comment by alculquicondor** [comment]
/lgtm
/approve
/hold
for two more runs

When retesting, leave a comment saying how many times it has passed already.

**Comment by kannon92** [comment]
First two runs have passed.
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-kind-ipv6

**Comment by kannon92** [comment]
three runs have passed.  

/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-kind-ipv6

/hold cancel
@alculquicondor has reviewed and approved but it looks like we need some other approvers on this one.
/cc @soltysh @pohly

**Comment by kannon92** [comment]
four runs have passed.

/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-kind-ipv6

**Comment by carlory** [comment]
@kannon92 I find the root cause and raise a pr to fix it. we may not need to change the existing test case. can you help to review https://github.com/kubernetes/kubernetes/pull/121051

**Comment by aojea** [comment]
> @kannon92 I find the root cause and raise a pr to fix it. we may not need change the existing test case. can you help to review #121051

/hold

this test started to flake at one point in time so we should find the root cause and not try to workaround it by adding retries, if those were not needed before is weird we need it now

**Comment by aojea** [comment]
for the record, in case we clearly identify a PR or commit that caused the flakiness but we didn't root cause the problem, it is better to revert the PR and ask the author of the PR to investigate the problem than fix forward, because we'll have the problem hidden in the codebase without noticing 

Anyway, thanks a lot for the great work detecting the problem and coming up with a fix

**Comment by kannon92** [comment]
@carlory found the root cause. It looks like it was a test framework issue and not any code changes to the job controller.

**Comment by kannon92** [comment]
@alculquicondor how do you want to proceed with this PR? It looks like it is not needed for the flake but I think it could be a good addition. It follows the other tests where we wait for a condition rather than listing pods.  But I can close it also.

**Comment by alculquicondor** [comment]
/remove-kind flake
/kind cleanup
/retitle Remove consistent check for job suspend

**Comment by alculquicondor** [comment]
/retitle Remove consistent check in test for job suspend

File: test/e2e/framework/job/wait.go
Comment (carlory) [inline test/e2e/framework/job/wait.go:73]: @kannon92 WaitFor + something  isn't a good pattern for new test cases, see https://github.com/kubernetes/kubernetes/issues/115164 for details. please consider using gomega.Eventually to handle this case, https://github.com/kubernetes/kubernetes/pull/119695 is a good example.

Comment (kannon92) [reply test/e2e/framework/job/wait.go:73]: Hmm.  I see.  I did not know that this was considered an antipattern.

We can close this PR as is if this is the better approach for testing.

Comment (carlory) [reply test/e2e/framework/job/wait.go:73]: /cc @pohly 
can you take a look at it?

Comment (mimowo) [reply test/e2e/framework/job/wait.go:73]: In this test we need some form of waiting as the Job controller needs time to add the condition, based on the  `.spec.suspend` field, which happens here: https://github.com/kubernetes/kubernetes/blob/3d1ac86013db78a270a5dedbdedc63027333aef4/pkg/controller/job/job_controller.go#L893-L900, and is later send to the api server.

Currently, this waiting is hidden inside `Consistently` which verifies the pods aren't created. IMO, it improves readability to have some explicit wait for the suspend field. So, I would suggest just adjust the PR to use `gomega.Eventually`. IIUC using `WaitForJobSuspend` or  `gomega.Eventually` is more of a technical detail.

Comment (pohly) [reply test/e2e/framework/job/wait.go:73]: > I would suggest just adjust the PR to use gomega.Eventually.

Agreed.

**Comment by mimowo** [comment]
I see the last change in this PR makes use of `gomega.Eventually`, which was the remaining remark.
/lgtm

**Comment by mimowo** [comment]
Please also update the description. In particular:
- replace  `Fixes #121013` with `Related to `#121013`.
- "What this PR does / why we need it:" section. I would suggest: "Eliminate potential race condition when checking if the job has Suspended condition. Current implementation is fragile because it takes time for the Job controller to add the condition. On the other hand, the waiting in test happens as a side effect of calling `Consistently` function which is used to verify pods aren't created for a while.

File: test/e2e/framework/job/wait.go
Comment (pohly) [inline test/e2e/framework/job/wait.go:75]: Here you are using Go code to check one attribute of a job. If this fails, the `job should be suspended` error becomes the failure reported by `gomega.Eventually`, with no additional information about what the actual status of the job is.

A better approach is to repeatedly get the job, then use gomega to check for the expected status. See https://github.com/kubernetes/kubernetes/blob/dc8b57d8a75764e649d641ac77c1539a574d9aed/test/e2e/framework/pod/wait.go#L221-L239 for an example.

File: test/e2e/framework/job/wait.go
Comment (pohly) [inline test/e2e/framework/job/wait.go:74]: When using gomega.Eventually, a failure is attributed to the helper, not the test, and the test cannot provide additional details. Better return an error, as explained in https://www.kubernetes.dev/blog/2023/04/12/e2e-testing-best-practices-reloaded/#polling-and-timeouts ("The E2E framework therefore uses a different approach").

**Review by pohly — CHANGES_REQUESTED** [review]

File: test/e2e/framework/job/wait.go
Comment (alculquicondor) [reply test/e2e/framework/job/wait.go:75]: But here we are not waiting for the job to exist (we rather know it does). We are waiting for it to be picked up by the reconciler.

Can we use `EventuallyWithOffset`?

Comment (pohly) [reply test/e2e/framework/job/wait.go:75]: > But here we are not waiting for the job to exist (we rather know it does). We are waiting for it to be picked up by the reconciler.

So that's what the custom matcher must check for.

If it's known that it exists, then the `framework.RetryNotFound` should not be used. Then `NotFound` will be treated as an error which aborts polling immediately.

> Can we use EventuallyWithOffset?

And use which offset? When using 1, the assumption is that `WaitForJobSuspend` is always called directly by tests. That's not necessarily true: it might also be called by some helper function in a test.

Comment (pohly) [reply test/e2e/framework/job/wait.go:75]: My suggestion is to do something like the `framework.Gomega` in `WaitForPodCondition` also in `WaitForJobSuspend`.

Comment (kannon92) [reply test/e2e/framework/job/wait.go]: I updated the PR

File: test/e2e/framework/job/wait.go
Comment (alculquicondor) [inline test/e2e/framework/job/wait.go:77]:
```suggestion
	return WaitForJobState(ctx, c, ns, jobName, "suspended", JobTimeout, func(job *batchv1.Job) (bool, error) {
```

?

`Condition` confused me a little, just because there is the `job.Status.Conditions` field. But IIUC, this is supposed to be more generic.

File: test/e2e/framework/job/wait.go
Comment (alculquicondor) [inline test/e2e/framework/job/wait.go:168]: what is the `1` for?

Comment (kannon92) [reply test/e2e/framework/job/wait.go:168]:
```
/*
Pretty prints the passed in object at the passed in indentation level.

Object recurses into deeply nested objects emitting pretty-printed representations of their components.

Modify format.MaxDepth to control how deep the recursion is allowed to go
Set format.UseStringerRepresentation to true to return object.GoString() or object.String() when available instead of
recursing into the object.

Set PrintContextObjects to true to print the content of objects implementing context.Context
*/
func Object(object interface{}, indentation uint) string {
	indent := strings.Repeat(Indent, int(indentation))
	value := reflect.ValueOf(object)
	commonRepresentation := ""
	if err, ok := object.(error); ok && !isNilValue(value) { // isNilValue check needed here to avoid nil deref due to boxed nil
		commonRepresentation += "\n" + IndentString(err.Error(), indentation) + "\n" + indent
	}
	return fmt.Sprintf("%s<%s>: %s%s", indent, formatType(value), commonRepresentation, formatValue(value, indentation))
}
```

File: test/e2e/framework/job/wait.go
Comment (kannon92) [reply test/e2e/framework/job/wait.go]: updated.

File: test/e2e/framework/job/wait.go
Comment (kannon92) [reply test/e2e/framework/job/wait.go]: Thank you.  That was a useful read for this.

File: test/e2e/framework/job/wait.go
Comment (pohly) [inline test/e2e/framework/job/wait.go:34]: As users of this packages might have to declare a variable of this type it makes sense to export it. Please also document the semantic of the return values.

Comment (pohly) [reply test/e2e/framework/job/wait.go:34]: I know, pod/wait.go doesn't do that either - that's from before my time (4969a05327be), I would have done it differently. But changing it in pod/wait.go would have broken quite a bit of code.

Speaking of differently: the prototype seems to come from the wait package. It allows four different combinations of return values, with some ambiguity what each combination means. Can we get the same effect with a single result, for example a simple failure string? Returning an empty string implies success.

Then the `conditionDesc` parameter in `WaitForJobState` can be dropped.

File: test/e2e/framework/job/wait.go
Comment (alculquicondor) [reply test/e2e/framework/job/wait.go]: `1` sounds useless then? we wouldn't see the conditions. But that's a nit.

File: test/e2e/framework/job/wait.go
Comment (alculquicondor) [inline test/e2e/framework/job/wait.go:37]:
```suggestion
type JobState func(job *batchv1.Job) (bool, error)
```

**Review by alculquicondor — COMMENTED** [review]
/approve
/hold for nit

File: test/e2e/framework/job/wait.go
Comment (kannon92) [reply test/e2e/framework/job/wait.go]: updated.

File: test/e2e/framework/job/wait.go
Comment (kannon92) [reply test/e2e/framework/job/wait.go]: updated.

**Comment by alculquicondor** [comment]
This PR is already approved, you'll likely need to rebase
https://github.com/kubernetes/kubernetes/pull/121300

**Review by pohly — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
